File: internal/controllers/image.go
// Copyright 2025 The wharf Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controllers

import (
	"context"
	"log"
	"net/http"
	"time"

	"github.com/docker/docker/api/types/image"
	"github.com/docker/docker/client"
	"github.com/docker/docker/errdefs"
	"github.com/gin-gonic/gin"
	"github.com/go-playground/validator/v10"

	dockerImage "github.com/wharf/wharf/pkg/image"
	"github.com/wharf/wharf/pkg/models"
)

func GetImages(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		images, err := dockerImage.GetAll(ctx, dockerClient)
		if err != nil {
			log.Println(err)
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(200, images)
	}
}

func PruneImages(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		ur, _ := c.Get("users")
		reqUser, _ := ur.(models.User)
		if reqUser.Permission == models.Read {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid premissions"})
			return
		}
		report, err := dockerImage.Prune(ctx, dockerClient)
		if err != nil {
			log.Println(err)
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, report)
	}
}

func RemoveImage(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		id := c.Param("id")
		var requestBody dockerImage.ImageRemoveRequest
		if err := c.BindJSON(&requestBody); err != nil {
			log.Println(err)
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		ur, _ := c.Get("user")
		reqUser, _ := ur.(models.User)
		if reqUser.Permission != models.Execute {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid permissions"})
			return
		}

		opts := image.RemoveOptions{
			Force:         false,
			PruneChildren: false,
		}

		if requestBody.Force != nil {
			opts.Force = *requestBody.Force
		}

		if requestBody.PruneChildren != nil {
			opts.PruneChildren = *requestBody.PruneChildren
		}
		report, err := dockerImage.Remove(ctx, dockerClient, id, opts)
		if err != nil {
			if errdefs.IsNotFound(err) {
				c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
				return
			}
			log.Println(err)
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, report)
	}
}

func TagImage(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		id := c.Param("id")
		ur, _ := c.Get("user")
		reqUser, _ := ur.(models.User)
		if reqUser.Permission == models.Read {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid permissions"})
			return
		}
		var reqBody dockerImage.ImageTagRequest
		if err := c.BindJSON(&reqBody); err != nil {
			log.Println(err)
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		validate := validator.New()
		if err := validate.Struct(reqBody); err != nil {
			log.Println(err)
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		err := dockerImage.Tag(ctx, dockerClient, id, reqBody.Tag)

		if err != nil {
			if errdefs.IsNotFound(err) {
				c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
				return
			}
			log.Println(err)
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, gin.H{"message": id + " tagged successfully"})

	}
}

File: client/src/api/image.ts
// Copyright 2025 The wharf Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import axios, { AxiosResponse } from 'axios';
import { DeleteResponse, Image, ImagesPruneReport } from '../models/image';
import { hostUrl } from '../utils/util';
import { MessageRes } from '../models/common';

const URL = hostUrl();

export const getAllImages = async (
  token: string
): Promise<AxiosResponse<Image[]>> => {
  const url = URL + '/api/protected/image/getAll';
  const res = await axios.get<Image[]>(url, {
    headers: {
      Token: token,
    },
  });
  return res;
};

export const pruneImages = async (
  token: string
): Promise<AxiosResponse<ImagesPruneReport>> => {
  const url = URL + '/api/protected/image/prune';
  const res = await axios.delete<ImagesPruneReport>(url, {
    headers: {
      Token: token,
    },
  });
  return res;
};

export const deleteImagge = async (
  token: string,
  id: string,
  force: boolean,
  pruneChildren: boolean
): Promise<AxiosResponse<DeleteResponse[]>> => {
  const url = URL + '/api/protected/image/remove/' + id;
  const res = await axios.delete<DeleteResponse[]>(url, {
    headers: {
      Token: token,
    },
    data: {
      force: force,
      pruneChildren: pruneChildren,
    },
  });
  return res;
};

export const tagImage = async (
  token: string,
  id: string,
  tag: string
): Promise<AxiosResponse<MessageRes>> => {
  const url = URL + '/api/protected/image/tag/' + id;
  const res = await axios.put<MessageRes>(
    url,
    {
      tag: tag,
    },
    {
      headers: {
        Token: token,
      },
    }
  );
  return res;
};
